internal/provisioner/plain/controllers/bundleinstance_controller.go

```diff
@@ -154,6 +156,32 @@ func (r *BundleInstanceReconciler) Reconcile(ctx context.Context, req ctrl.Reque
 			})
 			return ctrl.Result{}, err
 		}
+
+		/*
+			Check if the current object is a CRD. If it is, validate it to ensure safe and
+			non-destructive upgrades.
+
+			TODO (tylerslaton): Convert crd validation into an AdmissionWebhook
+
+			Eventually we want to convert this code segment into an webhook that validates
+			this logic agnostic of any one provisioner. There was some contention around.
+			Eventually it was settled on implenting this interface but should not be the end
+			product for this behavior.
+		*/
+		var unmarshalledCRD apiextensionsv1.CustomResourceDefinition
+		err = yaml.Unmarshal(jsonData, &unmarshalledCRD)
+		if err == nil && unmarshalledCRD.GroupVersionKind() == apiextensionsv1.SchemeGroupVersion.WithKind("CustomResourceDefinition") {
```

Comment:
This type check is a bit awkward -- I'm not sure but since we are unmarshaling to an `apiextensionsv1.CustomResourceDefinition` concrete type I imagine the GVK for that type will always be the v1 apiextensions GVK regardless of whether the underlying JSON represents a v1 CRD in reality. If it doesn't, then an error occurs (probably some field is not able to be unmarshaled successfully) which doesn't seem to be handled here. 

I would recommend as an intermediate step unmarshaling the jsonData into an unstructured.Unstructured object, which has a `GroupVersionKind()` method. From there we can do the GVK equality check and be more confident that the object we are handling is actually a CRD. It also shouldn't error since we are just converting JSON into another JSON-based type.

Reply:
Alternatively we can move this check up to L146 to do a `obj.GetObjectKind()` when we are handling the client.Object directly and handle the CRD from there..it makes the code a bit less clear though.